src/main/java/com/wordcount/WordCount.java

```diff
@@ -1,24 +1,27 @@
 package com.wordcount;
 
 import com.wordcount.controllers.WordCounterController;
-import com.wordcount.readers.ConsoleTextReader;
+import com.wordcount.readers.ConsoleReader;
 import com.wordcount.readers.StopWordsReader;
-import com.wordcount.readers.impl.ConsoleTextReaderImpl;
+import com.wordcount.readers.UserInputReader;
+import com.wordcount.readers.impl.ConsoleReaderImpl;
 import com.wordcount.readers.impl.StopWordsReaderImpl;
-import com.wordcount.writers.AnswerWriter;
+import com.wordcount.readers.impl.UserInputReaderImpl;
 import com.wordcount.writers.ConsoleWriter;
-import com.wordcount.writers.impl.AnswerWriterImpl;
+import com.wordcount.writers.UserOutputWriter;
 import com.wordcount.writers.impl.ConsoleWriterImpl;
+import com.wordcount.writers.impl.UserOutputWriterImpl;
 
 public class WordCount {
 
     public static void main(String[] args) {
-        ConsoleTextReader textReader = new ConsoleTextReaderImpl();
+        ConsoleReader consoleReader = new ConsoleReaderImpl();
+        UserInputReader inputReader = new UserInputReaderImpl(consoleReader);
```

Comment:
**Impl** suffix is added by me everywhere just because I used it a lot in my previous projects where we went  according the book Clean code where it is recommended to add **impl** to implementation. I can stop using it.

Reply:
I tried to improve design and added a couple of interfaces. You can see it here https://lucid.app/lucidchart/invitations/accept/2aa6722c-a3cc-4e81-b908-29e18eafb726
So, I will inject abstract IOCommunicator to Controller. By IOCommunicator I mean UI abstract layer.
Is it better?

Reply:
> **Impl** suffix is added by me everywhere just because I used it a lot in my previous projects where we went according the book Clean code where it is recommended to add **impl** to implementation. I can stop using it.

Some people like to [prefix](https://stackoverflow.com/questions/5816951/prefixing-interfaces-with-i) their abstractions (Intefaces) with an **I** (.NET community). Other people like to [postfix](https://stackoverflow.com/questions/541912/interface-naming-in-java) their implementations with **Impl** (Java community). Both hint at a lack of necessity for the abstraction since you only have one implementation or the abstraction is "just" a [header interface instead of a role interface](https://blog.shaunfinglas.co.uk/2015/10/header-interfaces-vs-role-interfaces.html). 

[Uncle Bob](http://cleancoder.com/products) (Clean Code book author) wants to make the point, that in doubt you should focus on the abstraction's name and therefore rather postfix the implementation with **Impl** instead of prefixing the abstraction with an **I**.

The point is, that both are not really a good idea. Abstractions should not reveal implementation details and implementations should be named after their details. For example: `UserRepository ur = new OracleDatabaseUserRepository()` (Domain Driven Design - [Repository Pattern](https://svatasimara.medium.com/domain-driven-design-part-5-repository-d5ad32b2e06f)). Another example would be: `UserInterface ui = new ConsoleUserInterface()`.